Redesign the graph editor for visual shaders

Geometror opened the pull request:
![grafik](https://github.com/godotengine/godot/assets/50084500/07e57556-1102-4efc-bb71-cba29a61db18)

~~Depends on #83785.~~ (merged)
~~Depends on #86158.~~ (merged)
~~Although it does not depend on #83510, it would make sense to merge that PR first.~~ (merged)

This PR aims to improve the VisualShader workflow especially when working with complex graphs where it's important that you're able

 to quickly parse the graph and get an overview about what each part does (at least roughly). As color is a very powerful instrument to guide your eyes and aid with comprehension of complex structures, we need to carefully apply it.

# Comparisons
Comparison using a complex node graph (thanks to @TheRensei for the shader which was very useful for testing/iterating):

Current:
![godot windows editor dev x86_64_lXxlqxNNZK](https://github.com/godotengine/godot/assets/50084500/1bc755e0-c3e4-44f4-84f8-95e82c233354)

This PR:
![jJVZVN8tEr](https://github.com/godotengine/godot/assets/50084500/89eb8251-a234-4b1a-80ff-901a5951731d)


# Detailed changes
- Use dotted grid pattern by default
- Refreshed VisualShader node design
	- Adjust margins to gain some "space to breathe"
	- Add dropshadow to text for better contrast/readability
- Color VisualShader nodes (header/titlebar) by category. Implements godotengine/godot-proposals#7745.
	- Category colors do not match exactly with the top ones in the "Add node" dialog since those sometimes aren't well suited for coloring the graph in a sensible manner. For certain nodes it might make sense to move them in the dialog as well.
		- e.g. inputs, parameters and constants all have the same color
		- e.g. VaryingGetter is colored like other input nodes and VaryingSetter is colored like the output node despite the fact that both nodes can be found in the "Special" category in the "Add node" dialog (tbh I would move them in the dialog too)
		- no common category (only contained DerivateFunc which is colored like the utility nodes) 
- Rework the connection type colors to improve the clarity of large graphs (and to match Blender where it's possible/reasonable, lowers "readjustment overhead" between Godot and Blender)
- Category and connection colors can be customized via editor settings
	- `Default` and `Legacy` (old connections colors + no distinct colors for categories) theme presets
	- `Custom` color theme allows setting each color manually
![Pasted image 20231117150123](https://github.com/godotengine/godot/assets/50084500/298a3c6f-c322-43d8-9b52-7ee2386b88fb)
- Optimization: Fix creating a new MSDF font theme for every node
- Set the minimum width of a graph node based on the preview size to avoid a resize when toggling the review

Since this is huge change we need to make sure that everything fits, so feel free to give it a try and let me know what you think and what could be improved. There are still some things to be discussed such as the color of Vector3/4 connections (maybe having two colors for Vector3, one for color related nodes and the other one for general vector nodes).

Thanks to everyone who gave such nice and valuable feedback during GodotCon.

_Production edit: closes godotengine/godot-roadmap#12_

Calinou commented:
This looks great! I wonder if the margin headings could be reduced though, as they look massive at the default zoom level currently. I understand this was done to make heading colors visible at low zoom levels, but maybe a different approach can be used to highlight these at low zoom levels instead.

Smaloki commented:
> I understand this was done to make heading colors visible at low zoom levels, but maybe a different approach can be used to highlight these at low zoom levels instead.

Maybe at a certain point while zooming out, the header colours could simply be extended to the entire node?

With a sufficiently low zoom factor, the text stops being readable anyway. At that point, the nodes might as well be represented by coloured rectangles. An instant switch without transition might be a little sudden though.

aaronfranke commented:
I see that this PR changes the colors for various types. The type colors should be kept in sync with the icon colors, so I would suggest either changing both or neither.

OctagonalHexy commented:
Will these changes break compatibility enough that I should wait before making custom shader nodes?

AThousandShips commented:
The `GraphEdit` class is experimental so it might change further before it's considered stable, you can use it as it is but you might need to rewrite your code if it changes

Geometror commented:
Rebased. The connection rim was split from this PR and is now part of https://github.com/godotengine/godot/pull/86158 (which this PR depends on).

Geometror committed:
Redesign the graph editor for visual shaders

Geometror commented:
Rebased and added a new category and color for particle nodes:
![Pasted image 20240209141820](https://github.com/godotengine/godot/assets/50084500/a0f65925-42fe-4e91-a4ca-83e84989c044)
This should be ready to review now :)

@aaronfranke 
I started changing the colors of the icons but then I realized that these icons are used all over the engine. Since the connection colors here can be changed (and are part of the VS theme) this didn't make much sense to me. I'd suggest we decide on this later and change them in a followup PR (IMO those icons need an overhaul anyway :)).

Geometror commented:
@Calinou The header heights are that big for another reason: More space for selecting/grabbing the nodes. Especially at high zoom levels (but not exclusively) this improves the flow since you don't need such a high click accuracy - reducing missclicks. Changing the height dynamically depending on the zoom levels might prove quite difficult and could cause all sorts of problems, so I'd rather not do this. (unfortunately I don't have any other ideas to highlight them at low zoom levels)
However, we could add an editor setting for the height I guess.

Calinou commented:
I guess it makes sense to keep the heading heights this way then.

aaronfranke commented:
@Geometror In that case I would suggest using the old colors, so that the colors stay in sync until such an icon redesign overhaul can be completed.

Geometror commented:
@aaronfranke Well, changing the connection colors is an integral part of this redesign, I tried the new header colors together with the old connection colors and it looked horrible. For example, for scalar values it's gray which isn't a saturated color on purpose since scalar connections are the most common type and therefore prevalent in most shader graphs. Having those colored (especially in the cyan used before) makes the graph way too colorful and hurts readability (since it interferes with the header colors). You're right, the color white is often used in scripting languages for control flow/sequencing, but on the other side, it is also used for data (or specifically scalar types) in Blender, Material Maker and the the Unreal material editor (Unity used colors, but very desaturated).  As this redesign is primarily inspired by Blender (to reduce fricition/cognitive overhead) I went with gray.
VisualScript is not part of the engine anymore, I don't think we should strive for consistency with something that's not even there (and has a different purpose), but rather go for consistency with equivalent tooling. 
Plus, we could still use white for control flow in Visual Scripting (extension) since it's way brighter than this gray.

Calinou reviewed (commented):
Tested locally (rebased on top of `master` fb10e67fefb85af3b8f5abd30db561588d8456c6), it works as expected. It looks great :slightly_smiling_face: 

One issue I noticed is that if you use the **Black (OLED)** editor theme preset, unselected nodes will blend with the background:

![image](https://github.com/godotengine/godot/assets/180032/54ead711-4ca2-4ef2-a654-d65612756256)

To combat this, there is a **Draw Extra Borders** editor setting that is automatically enabled when choosing this theme preset. When this setting is enabled, I suggest adding a gray 1-pixel border around unselected GraphNodes like we do on LineEdit, for instance.

Geometror commented:
@Calinou While adjusting the theme for the OLED preset, I realized that the contrast of the node borders is generally pretty low. So I played a bit with it: 

![image](https://github.com/godotengine/godot/assets/50084500/61f1846c-ec89-466a-a20b-eccda5dd30b9)

What do you think?

(I took the following two screenshots before adjusting the port offset)
Here's how it looks like with the OLED preset now:
![image](https://github.com/godotengine/godot/assets/50084500/94033c00-3c20-470e-9b6c-b989be23f0e3)
And the white theme was a bit broken since the dark font color doesn't work well with the header colors (which were choosen in a way that works for both light and dark themes), so I fixed that as well:
![image](https://github.com/godotengine/godot/assets/50084500/57f1ee4c-d7c1-48b3-aedb-9a4ce047cf06)

Geometror commented on doc/classes/EditorSettings.xml:
```diff
@@ -442,6 +442,65 @@
+			The color of a port/connection of sampler type.
+		</member>
+		<member name="editors/visual_editors/connection_colors/scalar_color" type="Color" setter="" getter="">
+			The color of a port/connection of scalar type (float, int, unsigned int).
```
In the docs the full name is used most of the time.

Geometror commented:
While at it, I improved the visibility of selected nodes. This was pretty bad at low zoom levels before.

![image](https://github.com/godotengine/godot/assets/50084500/f10e1224-c83a-4c04-bbca-9c403dfcdbb6)

![image](https://github.com/godotengine/godot/assets/50084500/6befffeb-d4b0-4963-bc94-0abab78098e2)

Mickeon commented on doc/classes/EditorSettings.xml:
```diff
@@ -445,6 +445,65 @@
+		<member name="editors/visual_editors/connection_colors/vector4_color" type="Color" setter="" getter="">
+			The color of a port/connection of Vector4 type.
+		</member>
+		<member name="editors/visual_editors/grid_pattern" type="int" setter="" getter="">
+		</member>
```
Something missing here. You'd know best what this is for.

akien-mga commented:
Thanks!

Jeangowhy commented:
Well done!

MangoDragonHub commented:
I don't know if this is in the new version but does this new version of the node system support comment grouping for chunks of nodes? It's just an organization thing. For example, you highlight a group of nodes for one part of the script and press a shortcut key (C), that puts that section in a comment bubble. Its easier to see what part of the shader is doing what.

Mickeon commented:
Do you mean something like this https://github.com/godotengine/godot/pull/88014 ?
